core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpRequest.java

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import static com.linecorp.armeria.common.util.Exceptions.throwIfFatal;
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.SplitHttpRequest;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.NoopSubscriber;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+import com.linecorp.armeria.unsafe.PooledObjects;
+
+import io.netty.util.concurrent.EventExecutor;
+
+public class DefaultSplitHttpRequest implements StreamMessage<HttpData>, SplitHttpRequest {
```

Comment:
I saw a lot of duplicate code in this class and `DefaultSplitHttpResponse`. I think we should dedup them.
How about adding:
```
public interface SplitHttpMessage {
    StreamMessage<HttpData> body();
    CompletableFuture<HttpHeaders> trailers();
}
```
and `SplitHttpResponse` and `SplitHttpRequest` extends it?

We can also do:
```java
public final class DefaultSplitHttpRequest extends AbstractSplitHttpMessage implements SplitHttpRequest {

    @Override
    public RequestHeaders headers() {
        return request.headers();
    }

    ...
    @Override
    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
                              SubscriptionOption... options) {
        // Need to use different BodySubscriber for request and response
    }
}

class BodySubscriber implements Subscriber Subscriber<HttpObject>, Subscription {
    ...
}

class SpliteResponseBodySubscriber extends BodySubscriber {
    
    private long pendingRequests = 1;

    private final HeadersFuture<ResponseHeaders> headersFuture;
    ...

    @Override
    public void onNext(HttpObject httpObject) {
        if (httpObject instanceof ResponseHeaders) {
            final ResponseHeaders headers = (ResponseHeaders) httpObject;
            final HttpStatus status = headers.status();
            if (status.isInformational()) {
                // Ignore informational headers
                upstream.request(1);
            } else {
                headersFuture.doComplete(headers);
            }
            return;
        }
        super.onNext(httpObject);
    }
    ...
}



```

Reply:
Thank you for the good comments. As you said, there was a lot of duplicate code, so I did a refactoring.
The refactored details are as follows.
- `SplitHttpMessage` interface has been added, and the existing `SplitHttpRequest` and `SplitHttpMessage` classes have been modified to implement this interface.
- `AbstractSplitHttpMessage` class has been added, and the existing `DefaultSplitHttpRequest` and `DefaultSplitHttpResponse` classes have been modified to extends this class.